src/pages/api/inboundMatching.ts

```diff
@@ -0,0 +1,175 @@
+import { NextApiRequest, NextApiResponse } from 'next';
+import { redirectTo, redirectToError, getUuidFromCookie, setCookieOnResponseObject, getDomain } from './apiUtils';
+import { BasicService } from '../../interfaces/index';
+import { Stop } from '../../data/auroradb';
+import { getMatchingFareStages, putStringInS3, UserFareStages } from '../../data/s3';
+import { isCookiesUUIDMatch, isSessionValid } from './service/validator';
+import { MATCHING_DATA_BUCKET_NAME, MATCHING_COOKIE } from '../../constants';
+
+interface FareZones {
+    name: string;
+    stops: Stop[];
+    prices: {
+        price: string;
+        fareZones: string[];
+    }[];
+}
+
+interface MatchingData {
+    type: string;
+    lineName: string;
+    nocCode: string;
+    operatorShortName: string;
+    inboundFareZones: FareZones[];
+    outboundFareZones: FareZones[];
+}
+
+interface MatchingFareZones {
+    [key: string]: {
+        name: string;
+        stops: Stop[];
+    };
+}
+
+export const putMatchingDataInS3 = async (data: MatchingData, uuid: string): Promise<void> => {
+    await putStringInS3(
+        MATCHING_DATA_BUCKET_NAME,
+        `${uuid}.json`,
+        JSON.stringify(data),
+        'application/json; charset=utf-8',
+    );
+};
+
+const getMatchingFareZonesFromForm = (req: NextApiRequest): MatchingFareZones => {
+    const matchingFareZones: MatchingFareZones = {};
+    const bodyValues: string[] = Object.values(req.body);
+
+    bodyValues.forEach((zoneString: string) => {
+        if (zoneString && typeof zoneString === 'string') {
+            const zone = JSON.parse(zoneString);
+
+            if (matchingFareZones[zone.stage]) {
+                matchingFareZones[zone.stage].stops.push(zone.stop);
+            } else {
+                matchingFareZones[zone.stage] = {
+                    name: zone.stage,
+                    stops: [zone.stop],
+                };
+            }
+        }
+    });
+
+    if (Object.keys(matchingFareZones).length === 0) {
+        throw new Error('No Stops allocated to fare stages');
+    }
+
+    return matchingFareZones;
+};
+
+const getMatchingJson = (
+    service: BasicService,
+    userFareStages: UserFareStages,
+    inboundMatchingFareZones: MatchingFareZones,
+    outboundMatchingFareZones: MatchingFareZones,
+): MatchingData => ({
+    ...service,
+    type: 'return',
+    inboundFareZones: userFareStages.fareStages
```

Comment:
Could we extract
```
userFareStages.fareStages
        .filter(userStage => inboundMatchingFareZones[userStage.stageName])
        .map(userStage => {
            const matchedZone = inboundMatchingFareZones[userStage.stageName];

            return {
                name: userStage.stageName,
                stops: matchedZone.stops.map((stop: Stop) => ({
                    ...stop,
                    qualifierName: '',
                })),
                prices: userStage.prices,
            };
        })
```
into a function and pass it the userFareStages and the matching fare zones since it's used for both inbound and outbound here?